### ISSUE TITLE
Make CLI exit code useful for detecting `tmt` errors

### ISSUE THREAD

**Issue by comps** [teemtee/tmt#2756]
Please change the reserved exit codes, as documented on, https://tmt.readthedocs.io/en/stable/code/autodocs/tmt.html#tmt.cli.TmtExitCode , to use values 2 or greater for test results, leaving 1 for internal TMT-specific errors, which should always be investigated by the infrastructure owner.

These would be ie.
* TMT bug causing a python traceback (which naturally exits with 1)
* invalid command line arguments
   * or CLI arguments with invalid values, ie. existing `tmt run -i` directory
* missing python module (causing `NameError`, see first point)
* inability to create a writable temporary directory
* etc.

and potentially even
* `discover` from a git URL returning HTTP errors
* other "setup-style" tasks

**Comment by LecrisUT** [comment]
Would require some coordination with downstream, maybe gating this by an env-variable `TMT_NEW_EXIT_CODES`. Some more reference [^1][^2].

It does seem more in line to have:
- `1` catch-all internal error, maybe including Python module import errors, etc.
- `2` syntax error either in cli or fmf files
- other codes still to be discussed

Pytest does not seem to use such convention, but it's also only half aligned with tmt's. Why it might make sense to align more with other standards is that tmt can be used outside of test wrapping, e.g. `tmt show`, `tmt run discover`, etc. 

[^1]: https://unix.stackexchange.com/a/254747
[^2]: https://docs.python.org/3/library/sys.html#sys.exit

**Comment by psss** [comment]
Yeah, I would also lean towards keeping the current exit code definition and rather [catching & redirecting](https://stackoverflow.com/questions/6720119/setting-exit-code-in-python-when-an-exception-is-raised) all errors to a proper exit code. Unless really necessary let's not break the backward compatibility.

**Comment by happz** [comment]
> Yeah, I would also lean towards keeping the current exit code definition and rather [catching & redirecting](https://stackoverflow.com/questions/6720119/setting-exit-code-in-python-when-an-exception-is-raised) all errors to a proper exit code. Unless really necessary let's not break the backward compatibility.

Isn't that already happening in https://github.com/teemtee/tmt/blob/main/tmt/__main__.py#L15? Everything except for `SystemExit` gets mapped to `2`, "Errors occured during test execution.", which might be incorrect as many errors may happen before executing any test (e.g. the invalid URL used in `discover`).

**Comment by comps** [comment]
To clarify: My use case, essentially, is to catch any errors from an infrastructure PoV. I don't care if the tests reported (using custom format or not) pass/fail/error/skip/whatever.

I just want to know (somehow, ideally exit code, but doesn't have to be) that TMT failed to execute tests or retrieve their results.

Currently, I have to grep TMT's output for things like
* `^Error:`
* `^    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~$`
* `^plan failed$`
* etc.

**Comment by LecrisUT** [comment]
Does using exit code `2` not work in this case? Any non-test errors should be mapped to it

**Comment by comps** [comment]
> Does using exit code 2 not work in this case? Any non-test errors should be mapped to it

The problem is that any tests which report `error` (due to their internal logic, or just because they want to via `result:custom`) will also result in exit code 2.

If this was moved to some other exit code, then I could indeed use 2.

**Comment by LecrisUT** [comment]
Well, yes, but isn't that intended to be the case? `error` could mean that e.g. `beakerlib` was not installed, or anything to indicate that the testing infrastructure is not well prepared.

**Comment by comps** [comment]
The point is that `error` is a completely valid result status, and we use it (in our python non-Beakerlib suite) for several cases, ie.
* a failure-waiving rule matched a pass
* some timeouts inside a test were exceeded
* installing a nested VM failed on Anaconda
* a service (ie. `osbuild-composer`) failed to restart after the test made configuration changes to it
* etc.

We define, essentially,
* `pass` = test managed to run the expected scenarios and verify that they work
* `fail` = test managed to run the expected scenarios, verify that its results were valid (nothing test-related was broken) and the tested functionality failed
* `error` = a problem happened that prevented the test from testing the functionality

This means if we see a `fail`, it's almost always a new bug.

You can't rely on a test using `error` to mean an infrastructure error.

**Comment by psss** [comment]
> We define, essentially,

Hm, seems there's still some confusion. Let's sync on the definitions. From the [test results](https://tmt.readthedocs.io/en/stable/spec/results.html#) and [exit codes](https://tmt.readthedocs.io/en/stable/overview.html#exit-codes) specification:

> * `pass` = test managed to run the expected scenarios and verify that they work

* Test: Test execution successfully finished and passed.
* Run: At least one test passed, there was no fail, warn or error. This is exit code 0.

Both seem completely aligned.

> * `fail` = test managed to run the expected scenarios, verify that its results were valid (nothing test-related was broken) and the tested functionality failed

* Test: Test execution successfully finished and failed.
* Run: There was a fail or warn identified, but no error. Exit code 1.

Meaning tests completed an revealed a real problem, a warning might have occured.
Isn't this also aligned with what you are looking for?

> * `error` = a problem happened that prevented the test from testing the functionality

* Test: Undefined problem encountered during test execution. Human inspection is needed to investigate whether it was a test bug, infrastructure error or a real test failure.
* Run: Errors occurred during test execution. Exit code 2.

In short, something went wrong, need to investigate.

> This means if we see a `fail`, it's almost always a new bug.
> You can't rely on a test using `error` to mean an infrastructure error.

So, how I understand it:

* You can rely that everything (and at least something passed) if exit code is 0
* Test have successfully finished and revealed a real problem if exit cod is 1

If there has been an `error` then something went wrong and we cannot rely on the results but need to investigate. Do you suggest that we somehow should be able to "rely" on an error?

**Comment by comps** [comment]
The point of this GH issue was to try to separate tmt-reported problems from test-reported problems. And if a test can report `error` as a test result (which it can, so I considered it part of the result API accessible to tests), that's a test-reported problem.

In that case, `tmt` should exit with an exit code (ie. `2`) distinct from other tmt-reported problems (ie. exit `1`) when it tracebacks, or when it fails to upload tests, or when it fails to provision a guest, or when it fails to connect to a guest to download logs, etc.

Ideally, this would be some state reported outside of the `result` in `results.yaml` (perhaps as a `check`?) to not confuse it with test results, or perhaps as a new result, `infra` or something.

But since we don't have that, I filed this issue to at least distinguish between `error` reported by tmt (that our automation should probably auto-rerun, because it's likely a transient infra issue) and `error` reported by the test (because it's a valid test result according to the API exposed by tmt).

**Comment by psss** [comment]
Ok, thanks for clarification. So the core of the issue is to separate:

* test `error` which is reported by the test itself as part of the `results.yaml`
* any other `error` which prevents the test to be even executed

Then checking the `plan/execute/results.yaml` file for `result: error` could be used as a relatively simple workaround until a dedicated exit code is implemented.

> We could add a new exit code, to be used for tmt's internal errors, but at least some of them are outside of our current control - Click will raise `click.exceptions.NoSuchOption` and turn it into `SystemExit(2)` we do not intercept...

As @happz mentioned above, there might be some cases we currently do not handle, like invalid options identified by `click`. But even for those we could add `except SystemExit` and convert to a different exit code.

So the last question would be how the mapping should be, I'd probably suggest to keep exit codes which cover scenarios that the test itself was successfully executed close together:

    we were able to execute the test code
    0 ... pass
    1 ... fail
    2 ... error

    something weird or unexpected happened
    3 ... no results
    4 ... everything skipped
    5 ... guest prepare, network issue, wrong repo...

Would that make sense? Or shall we rather introduce some "section" of exit codes?

    10 ... unspecified error
    11 ... discover step problem
    12 ... guest provision failure
    13 ... error during prepare

We could start with just `10` as an unspecified error and extend with more specific codes later if needed.

**Comment by psss** [comment]
Summary from the triage meeting: The exit codes deserve to be improved, not necessarily adding new ones or too many new ones. Added to the backlog. Possibly relevant issue:

* #3888
